File: solutions/julia/game-of-life/1/game-of-life.jl
function gameoflife(matrix)
    length = size(matrix,1)
    width = size(matrix,2)
    if 0==length || 0==width
        return reshape([],0,2)
    end
    result = Array{Int64}(undef, length, width)
    for i in 1:length
        for j in 1:width
            sum = 0
            sum  += (i>1 && j>1) ? matrix[i-1, j-1] : 0    
            sum  += (i>1) ? matrix[i-1, j] : 0
            sum  += (i>1 && j+1<=width) ? matrix[i-1, j+1] : 0
            sum  += (j>1) ? matrix[i, j-1] : 0
            sum  += (j+1<=width) ? matrix[i, j+1] : 0
            sum  += (i+1<=length && j>1) ? matrix[i+1, j-1] : 0
            sum  += (i+1<=length) ? matrix[i+1, j] : 0
            sum  += (i+1<=length && j+1<=width) ? matrix[i+1, j+1] : 0
            result[i, j] = 
                ((3==sum) || (1== matrix[i, j] && 2==sum)) ? 1 : 0
        end
    end
    result
end

File: solutions/dart/game-of-life/1/lib/game_of_life.dart
class GameOfLife {
  List<List<int>> _matrix;

  GameOfLife(matrixIn) : _matrix = matrixIn;
  
  List<List<int>> matrix() => _matrix;
  

  void tick(){
    if (_matrix.isEmpty)
      return;
    final rows  = _matrix.length;
    final cols =  _matrix[0].length;
    final nextStep = List.generate(rows, 
                          (_) => List.filled(cols, 0));  
    for(var i=0; i<rows; i++){
      for(var j=0; j<cols; j++){
        var sum = countNeighbors(i,j);
         nextStep[i][j] = ((3==sum) || 
                           (1== _matrix[i][j] && 2==sum)) ?
                             1 : 0;   
      }
    }
    _matrix = nextStep;    
  }

  int countNeighbors(int i, int j){
    final rows  = _matrix.length;
    final cols =  _matrix[0].length;
    int sum = 0;
    sum  += (i>0 && j>0) ? _matrix[i-1][j-1] : 0;
    sum  += (i>0) ? _matrix[i-1][j]: 0; 
    sum += (i>0 && j+1<cols) ? _matrix[i-1][j+1] : 0;
    sum += (j>0) ? _matrix[i][j-1] : 0;
    sum += (j+1<cols) ? _matrix[i][j+1] : 0;
    sum  += (i+1<rows && j>0) ? _matrix[i+1][j-1] : 0;
    sum  += (i+1<rows) ? _matrix[i+1][j]: 0;
    sum += (i+1<rows && j+1<cols) ? _matrix[i+1][j+1] : 0;
    return sum;
  }

  
}
